### DISCUSSION TITLE
StyLua nvim plugin

### DISCUSSION THREAD

**Discussion by ckipp01** [JohnnyMorganz/StyLua#93] [Show and tell]
🙏🏼 huge thanks for StyLua. I've been wanting exactly this for quite some time, so when I landed on this I was quite happy. I was really missing an easy way to within Nvim to just format my file in a sensible way. Just in case anyone else is in this same position, I ended up throwing what I had locally into a plugin, [stylua-nvim](https://github.com/ckipp01/stylua-nvim). It doesn't do a ton, but does pretty much what you'd expect it to. I do plan on supporting range formatting with it as well.

**Comment by JohnnyMorganz** [comment]
That looks really cool, thanks for doing it!

I saw that you're handling errors through a bit of a hacky method, since it's not provided in a great way. Do you reckon I should provide some sort of toggle to output it in something like JSON? May be easier for you to parse

**Reply by ckipp01** [reply]
Ha, yes it's a bit hacky the way it is. Having some sort of structured output like JSON for that would definitely be way cleaner, yes. Maybe just hide it behind a flag or something?

**Reply by JohnnyMorganz** [reply]
Added in https://github.com/JohnnyMorganz/StyLua/pull/455, sorry for taking so long! (can't believe its already been over a year!😄)

**Reply by telemachus** [reply]
Thanks so much!

**Comment by telemachus** [comment]
Is this still under consideration? I was just coming to ask whether you might consider a patch to output errors in a more structured way in order to handle them more neatly in {neo,}vim.

The dream, I suppose, would be a flag like `rg`'s `--vimgrep` that prints errors to `stderr` in exactly the format that vim wants (`filename:line-number:column-number:error-message`), but JSON would obviously be useful for more than just {neo,}vim.

**Reply by JohnnyMorganz** [reply]
Yeah this is definitely still something I want to do, just never got round to it.

We now have a `--output-format=json` flag, but this currently outputs diffs in json when using `--check`, not errors.

We could update the `env-logger` builder to output json for general errors in https://github.com/JohnnyMorganz/StyLua/blob/master/src/cli/main.rs#L327-L353, but we would also want to special case error handling for parse errors (so that we can return data about the location) at https://github.com/JohnnyMorganz/StyLua/blob/master/src/cli/main.rs#L79-L80 and https://github.com/JohnnyMorganz/StyLua/blob/master/src/cli/main.rs#L118-L119

I was thinking of an error structure something like
```json5
{
    "type": "error", // or "warn"/"info"/"debug" etc. and "parse_error"
    "message": "XYZ"
}
```
and if `type == "parse_error"` we also have a `filename` and `location` field like
```json5
{
    "type": "parse_error",
    "message": "XYZ",
    "filename": "src/test.lua",
    "location": {
        "start": 0, // byte offset from start of file
        "start_line": 0,
        "start_column": 0,
        "end": 0, // byte offsets from start of file
        "end_line": 0,
        "end_column": 0,
    }
}
```

Would this suit your usecases?

**Reply by telemachus** [reply]
> Would this suit your usecases?

Thanks for responding so quickly.  Yes, absolutely, that would suit what I'm thinking of. The key items for vim and neovim are the filename, the error message itself, and the line and column of where the error starts (and optionally the line and column where the error ends). It should be easy enough to extract this information out of JSON returned by stylua.
